tokio/src/io/stdio_common.rs

```diff
@@ -0,0 +1,118 @@
+//! Contains utilities for stdout and stderr.
+use crate::io::AsyncWrite;
+use std::pin::Pin;
+use std::task::{Context, Poll};
+/// # Windows
+/// AsyncWrite adapter that finds last char boundary in given buffer and does not write the rest.
+/// That's why, wrapped writer will always receive well-formed utf-8 bytes.
+/// # Other platforms
+/// passes data to `inner` as is
+#[derive(Debug)]
+pub(crate) struct SplitByUtf8BoundaryIfWindows<W> {
+    inner: W,
+}
+
+impl<W> SplitByUtf8BoundaryIfWindows<W> {
+    pub(crate) fn new(inner: W) -> Self {
+        Self { inner }
+    }
+}
+
+impl<W> crate::io::AsyncWrite for SplitByUtf8BoundaryIfWindows<W>
+where
+    W: AsyncWrite + Unpin,
+{
+    fn poll_write(
+        mut self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        buf: &[u8],
+    ) -> Poll<Result<usize, std::io::Error>> {
+        // following two ifs are enabled only on windows targets, because
+        // on other targets we do not have problems with incomplete utf8 chars
+
+        // ensure buffer is not longer than MAX_BUF
+        #[cfg(any(target_os = "windows", test))]
+        let buf = if buf.len() > crate::io::blocking::MAX_BUF {
+            &buf[..crate::io::blocking::MAX_BUF]
+        } else {
+            buf
+        };
+        // now remove possible trailing incomplete character
+        #[cfg(any(target_os = "windows", test))]
+        let buf = match std::str::from_utf8(buf) {
+            // `buf` is already utf-8, no need to trim it
+            Ok(_) => buf,
+            Err(err) if err.valid_up_to() == 0 => {
+                return Poll::Ready(Err(std::io::Error::new(
+                    std::io::ErrorKind::InvalidInput,
+                    "provided buffer does not contain utf-8 data",
+                )));
```

Comment:
The original issue says:

> So, windows stdout as a console only accepts utf-16 characters. Stdlib's stdout detects whether the Stdout is a console in Windows, and then it assumes the byte buffer is encoded in utf-8 and then converts it to utf-16 so it can be printed.

So `stdout` might not be a console, in which case non-utf8 is ok. If the data isn't utf-8, I think we should just try to forward it anyway.

Reply:
I changed trimming logic.
Now if buffer has utf8 error at start, or more that 8 bytes would be skipped, we assume that caller really wants to print non-utf8 data.

Reply:
This seems rather complicated. Why not just set a boolean flag if you've already printed non utf-8 data?

Reply:
I think such a flag can lead to following scenario:
1) User creates a `tokio::io::Stdout` instance.
2) User passes this instance to a library. That library tries to write some binary data, gets an error and ignores it. Wrapper observes it and sets flag to true.
3) Now user tries to write long utf8 string. However, since flag is set, wrapper no longer trims buffer, and write operation fails.

I.e. one binary write "poisons" Stdout instance, and following legitimate "text" writes will sometimes fail.

Reply:
I read the code again. You've convinced me.